build-info-extractor-pip/src/main/java/org/jfrog/build/extractor/pip/extractor/PipBuildInfoExtractor.java

```diff
@@ -57,58 +59,126 @@ Build extract(ArtifactoryDependenciesClient client, String repository, String in
      */
     Map<String, Dependency> buildDependenciesMap(Map<String, String> downloadedDependencies, ArtifactoryDependenciesClient client, String repository, Path executionPath, Log logger) throws IOException {
         Map<String, Dependency> dependenciesMap = new HashMap<>();
-        Set<String> missingDeps = new HashSet<>();
         DependenciesCache dependenciesCache = DependenciesCache.getProjectDependenciesCache(executionPath, logger);
 
+        Map<String, String> getFromArtifactoryMap = new HashMap<>();
         for (String pkgName : downloadedDependencies.keySet()) {
             String fileName = downloadedDependencies.get(pkgName);
-            Dependency dependency = null;
             if (StringUtils.isNotBlank(fileName)) {
-                // Get dependency info from Artifactory - may throw IOException.
-                dependency = getPackageInfoFromArtifactory(fileName, repository, client, logger);
-            } else {
-                // Get dependency info from cache.
-                if (dependenciesCache != null) {
-                    dependency = dependenciesCache.getDependency(pkgName);
-                }
-            }
-
-            if (dependency == null) {
-                // Add missing dependency.
-                missingDeps.add(pkgName);
+                // Dependency downloaded from Artifactory.
+                getFromArtifactoryMap.put(fileName, pkgName);
                 continue;
             }
+            // Dependency wasn't downloaded in this execution, get dependency info from cache.
+            if (dependenciesCache != null && dependenciesCache.getDependency(pkgName) != null) {
+                dependenciesMap.put(pkgName, dependenciesCache.getDependency(pkgName));
+            }
+        }
 
-            dependenciesMap.put(pkgName, dependency);
+        // Get dependencies from Artifactory.
+        Map<String, Dependency> dependenciesFromArtifactory = getDependenciesFromArtifactory(getFromArtifactoryMap, repository, client, logger);
+        if (dependenciesFromArtifactory != null) {
+            dependenciesMap.putAll(dependenciesFromArtifactory);
         }
 
+        // Prompt missing dependencies.
+        Set<String> missingDeps = downloadedDependencies.keySet().stream()
+                .filter(x -> !dependenciesMap.containsKey(x))
+                .collect(Collectors.toSet());
         promptMissingDeps(missingDeps, logger);
+
         DependenciesCache.updateDependenciesCache(dependenciesMap, executionPath);
         return dependenciesMap;
     }
 
-    Dependency getPackageInfoFromArtifactory(String fileName, String repository, ArtifactoryDependenciesClient client, Log logger) throws IOException {
-        String aql = String.format(PIP_AQL_FORMAT, repository, fileName);
-        AqlSearchResult searchResult;
+    /**
+     * Fetch required files-data from Artifactory and create its corresponding Dependency object.
```

Comment:
Let's also explain how this method fetches the epedencies from Artifactory, and which of the arguments it gets are meant to be used outside this method, after it finishes.